### PR TITLE
Add script to create kOps cluster and run tests

### PR DESCRIPTION
**Pull request by jaydeokar**
### Issue

N/A

### Description 
This script add the ability to spin up a Kops cluster and run the test suite. 

### Checklist
- [ ] Added tests that cover your change (if possible)
- [ ] Added/modified documentation as required (such as the `README.md`, or the `docs` directory)
- [X] Manually tested
- [x] Made sure the title of the PR is a good description that can go into the release notes

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [ ] Refactored something and made the world a better place :star2:

### CODE REVIEW

File: scripts/lib/kops.sh [hunk 1]
```diff
@@ -0,0 +1,49 @@
+create_kops_cluster(){
```
Comment (kishorj) [inline scripts/lib/kops.sh:1]: kops::create_cluster

File: scripts/lib/kops.sh [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    printf "Using ${KOPS_S3_BUCKET} as kops state store \n"
```
Comment (kishorj) [inline scripts/lib/kops.sh:3]: echo instead of printf

File: scripts/lib/kops.sh [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    printf "Starting cluster creation: ${CLUSTER_NAME} \n"
+    "${KOPS_BIN}/kops" update cluster --name "${CLUSTER_NAME}" --yes --admin
+
+    printf "Waiting for cluster to be up: ${CLUSTER_NAME} \n"
+    "${KOPS_BIN}/kops" validate cluster --wait 15m
+
+    printf "Cluster creation complete \n"
+}
+
+delete_kops_cluster(){
```
Comment (kishorj) [inline scripts/lib/kops.sh:45]: kops::delete_cluster

File: scripts/run_kops_test.sh [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    fi
+  fi
+
+  if ! delete_kops_cluster; then
+    printf "Delete cluster failed.. \n"
+  fi
+}
+
+: "${TEST_ID:=$RANDOM}"
+: "${REGION:='us-west-2'}"
```
Comment (kishorj) [inline scripts/run_kops_test.sh:35]: You could use :- to assign default value value only if it is missing. This provides a useful way to override the defaults when you invoke the scripts. For example,
REGION=${REGION:-"us-west-2"}

File: scripts/run_kops_test.sh [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+: "${SKIP_MAKE_TEST_BINARIES:=false}"
+: "${EXTRA_GINKGO_FLAGS:=''}"
+: "${FOCUS:=''}"
+: "${PRINT_CONTROLLER_LOGS:=false}"
+
+SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+GINKGO_TEST_BUILD="$SCRIPT_DIR/../test/build"
+CONTROLLER_IAM_POLICY_FILE="$(dirname "${BASH_SOURCE[0]}")/../docs/install/iam_policy.json"
+CONTROLLER_IAM_POLICY_NAME="aws-lb-controller-kops-${TEST_ID}"
+KOPS_S3_BUCKET="kops-alb-eks-${AWS_ACCOUNT_ID}"
```
Comment (kishorj) [inline scripts/run_kops_test.sh:51]: Is AWS_ACCOUNT_ID always going to be set? If not, you might have to do something like 
  AWS_ACCOUNT_ID=$(aws sts get-caller-identity --region ${REGION} --query Account --output text)

File: scripts/run_kops_test.sh [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+KOPS_STATE_STORE=s3://$KOPS_S3_BUCKET
+CLUSTER_NAME="kops-alb-test-cluster-${TEST_ID}.k8s.local"
+KOPS_BIN="~/kops_bin"
+SSH_KEYS=~/.ssh/kopstempkey
+USE_EKSCTL=false
+policy_arn=""
```
Comment (kishorj) [inline scripts/run_kops_test.sh:57]: use uppercase for variable names

File: scripts/run_kops_test.sh [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+cleanup(){
+
+  exitcode=$?
+  if [[ ${exitcode} -gt 0 ]]; then
+    printf "Error occurred. Attempting to clean setup.. \n"
+  else
+    printf "Tests finished. Attempting to clean setup.. \n"
+  fi
+
+  if ! uninstall_aws_lb_controller; then
```
Comment (kishorj) [inline scripts/run_kops_test.sh:15]: did you verify further cleanup steps will get executed if this step fails?

File: scripts/run_kops_test.sh [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+elif [[ $REGION == "cn-northwest-1" ]];then
+  image="961992271922.dkr.ecr.cn-northwest-1.amazonaws.com.cn/amazon/aws-load-balancer-controller"
+else
+  image="602401143452.dkr.ecr.us-west-2.amazonaws.com/amazon/aws-load-balancer-controller"
+fi
+
+trap cleanup EXIT
+
+# Step 1: Create kOps cluster and wait for it to complete
+create_kops_cluster
```
Comment (kishorj) [inline scripts/run_kops_test.sh:79]: Please put this in a function, for example main, and invoke the main function

File: scripts/lib/install.sh [hunk 9]
```diff
@@ -0,0 +1,63 @@
+install_aws_lb_controller() {
```
Comment (kishorj) [inline scripts/lib/install.sh:1]: Is this not handled by the ginkgo tests?

Comment (kishorj) [inline scripts/lib/install.sh:1]: This file involves helm chart installation/deletion. Maybe rename it to helm.sh. Then the functions like:
helm::install...

File: scripts/lib/common.sh [hunk 10]
```diff
@@ -32,15 +34,18 @@ if [[ $REGION == "cn-north-1" || $REGION == "cn-northwest-1" ]]; then
...
-  CLUSTER_INFO=$($GET_CLUSTER_INFO_CMD --endpoint $ENDPOINT)
+  if [[ -z "${ENDPOINT}" ]]; then
+    CLUSTER_INFO=$($GET_CLUSTER_INFO_CMD)
+  else
+    CLUSTER_INFO=$($GET_CLUSTER_INFO_CMD --endpoint $ENDPOINT)
+  fi
+
+  VPC_ID=$(echo $CLUSTER_INFO | jq -r '.cluster.resourcesVpcConfig.vpcId')
+
+  echo "VPC ID: $VPC_ID"
```
Comment (kishorj) [inline scripts/lib/common.sh:48]: Are these not required for kops?

File: scripts/run_kops_test.sh (hunk 7 above)
Comment (jaydeokar) [reply scripts/run_kops_test.sh]: Yes, the if statement catches the error. We are logging with echo and moving forward with other cleanup execution

File: scripts/lib/common.sh (hunk 10 above)
Comment (jaydeokar) [reply scripts/lib/common.sh:48]: I didn't quite catch that.. VPC_ID is required by ginkgo tests which we pass as a parameter

File: scripts/lib/install.sh (hunk 9 above)
Comment (jaydeokar) [reply scripts/lib/install.sh]: I don't think ginkgo installs it by default. It's similar to how we test using eksctl (e2e test)

File: scripts/run_kops_test.sh [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+main(){
+
+  # Step 1: Create kOps cluster and wait for it to complete
+  kops::create_kops_cluster
+
+  # Step 2: Create policy and attach to nodes role created by kOps
+  POLICY_ARN=$(iam::create_policy "${CONTROLLER_IAM_POLICY_NAME}" "${CONTROLLER_IAM_POLICY_FILE}" "${REGION}")
+  iam::attach_policy_to_role "${POLICY_ARN}" "nodes.${CLUSTER_NAME}"
+
+  # Step 3: Set hop limit to 2 for IMDSv2
```
Comment (kishorj) [inline scripts/run_kops_test.sh:88]: You've specified the vpc-id and the region during helm installation, IMDS need not be enabled. Did you run into any errors?

File: scripts/run_kops_test.sh [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+: "${KOPS_BINARY_VERSION:=v1.25.3}"
+: "${KUBERNETES_VERSION:=1.26.0}"
+: "${NODE_SIZE:=c5.xlarge}"
+: "${NODE_COUNT:=3}"
+: "${OS_OVERRIDE:=linux}"
+: "${ENDPOINT:=""}"
+: "${SKIP_MAKE_TEST_BINARIES:=false}"
+: "${EXTRA_GINKGO_FLAGS:=''}"
+: "${FOCUS:=''}"
+: "${PRINT_CONTROLLER_LOGS:=false}"
```
Comment (kishorj) [inline scripts/run_kops_test.sh:45]: lets print controller logs by default

File: scripts/lib/kops.sh [hunk 13]
```diff
@@ -0,0 +1,49 @@
+kops::create_kops_cluster(){
```
Comment (kishorj) [inline scripts/lib/kops.sh:1]: kops::create_cluster

File: scripts/run_kops_test.sh (hunk 11 above)
Comment (jaydeokar) [reply scripts/run_kops_test.sh:88]: Yeah if I don't set hop limit to 2, this is what we see in the logs
```
"NoCredentialProviders: no valid providers in chain. Deprecated.\n\tFor verbose messaging see aws.Config.CredentialsChainVerboseErrors
```

Comment (jaydeokar) [reply scripts/run_kops_test.sh:88]: https://github.com/kubernetes-sigs/aws-load-balancer-controller/issues/1747#issuecomment-752215264
It looked similar to this

Comment (kishorj) [reply scripts/run_kops_test.sh:88]: got it. We are not using IRSA here, thats why.

**Pull request closed without merging**